File: android/Showcase/src/main/java/com/spruceid/mobilesdkexample/wallet/Draft18OID4VPSupport.kt
package com.spruceid.mobilesdkexample.wallet

import androidx.compose.foundation.Image
import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.navigationBarsPadding
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.foundation.verticalScroll
import androidx.compose.material3.Button
import androidx.compose.material3.ButtonDefaults
import androidx.compose.material3.Checkbox
import androidx.compose.material3.CheckboxDefaults
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableIntStateOf
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.res.painterResource
import androidx.compose.ui.res.stringResource
import androidx.compose.ui.text.ParagraphStyle
import androidx.compose.ui.text.SpanStyle
import androidx.compose.ui.text.buildAnnotatedString
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.text.style.TextIndent
import androidx.compose.ui.text.withStyle
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import com.spruceid.mobile.sdk.KeyManager
import com.spruceid.mobile.sdk.rs.DidMethod
import com.spruceid.mobile.sdk.rs.DidMethodUtils
import com.spruceid.mobile.sdk.rs.Draft18PresentableCredential
import com.spruceid.mobile.sdk.rs.Draft18PresentationSigner
import com.spruceid.mobile.sdk.rs.Draft18RequestedField
import com.spruceid.mobilesdkexample.DEFAULT_SIGNING_KEY_ID
import com.spruceid.mobilesdkexample.R
import com.spruceid.mobilesdkexample.ui.theme.ColorBase300
import com.spruceid.mobilesdkexample.ui.theme.ColorBase50
import com.spruceid.mobilesdkexample.ui.theme.ColorBlue600
import com.spruceid.mobilesdkexample.ui.theme.ColorEmerald900
import com.spruceid.mobilesdkexample.ui.theme.ColorStone300
import com.spruceid.mobilesdkexample.ui.theme.ColorStone600
import com.spruceid.mobilesdkexample.ui.theme.ColorStone950
import com.spruceid.mobilesdkexample.ui.theme.Inter
import com.spruceid.mobilesdkexample.utils.credentialTypeDisplayName
import com.spruceid.mobilesdkexample.utils.removeUnderscores
import com.spruceid.mobilesdkexample.utils.splitCamelCase
import org.json.JSONObject

class Draft18Signer(keyId: String?) : Draft18PresentationSigner {
    private val keyId = keyId ?: DEFAULT_SIGNING_KEY_ID
    private val keyManager = KeyManager()
    private var jwk: String
    private val didJwk = DidMethodUtils(DidMethod.JWK)

    init {
        if (!keyManager.keyExists(this.keyId)) {
            keyManager.generateSigningKey(id = this.keyId)
        }
        this.jwk = keyManager.getJwk(this.keyId)?.toString()
            ?: throw IllegalArgumentException("Invalid kid")
    }

    override suspend fun sign(payload: ByteArray): ByteArray {
        return keyManager.signPayload(keyId, payload)
            ?: throw IllegalStateException("Failed to sign payload")
    }

    override fun algorithm(): String {
        val json = JSONObject(jwk)
        return try {
            json.getString("alg")
        } catch (_: Exception) {
            "ES256"
        }
    }

    override suspend fun verificationMethod(): String {
        return didJwk.vmFromJwk(jwk)
    }

    override fun did(): String {
        return didJwk.didFromJwk(jwk)
    }

    override fun jwk(): String {
        return jwk
    }

    override fun cryptosuite(): String {
        return "ecdsa-rdfc-2019"
    }
}

data class Draft18CredentialRequirement(
    val descriptorId: String,
    val displayName: String,
    val credentials: List<Draft18PresentableCredential>,
)

@Composable
fun Draft18DataFieldSelector(
    selectedCredential: Draft18PresentableCredential,
    requestedFields: List<Draft18RequestedField>,
    currentIndex: Int = 0,
    totalCount: Int = 1,
    onContinue: (selectedFields: List<String>) -> Unit,
    onCancel: () -> Unit,
    allClaims: JSONObject = JSONObject()
) {
    var selectedFields by remember(currentIndex) {
        mutableStateOf(requestedFields.filter { it.required() }.map { it.path() })
    }
    val paragraphStyle = ParagraphStyle(textIndent = TextIndent(restLine = 12.sp))
    val supportsSelectiveDisclosure = selectedCredential.selectiveDisclosable()
    val hasMoreCredentials = currentIndex + 1 < totalCount

    Column(
        modifier = Modifier
            .fillMaxWidth()
            .padding(horizontal = 24.dp)
            .padding(top = 48.dp)
    ) {
        if (totalCount > 1) {
            Text(
                text = "Credential ${currentIndex + 1} of $totalCount",
                fontFamily = Inter,
                fontWeight = FontWeight.Medium,
                fontSize = 14.sp,
                color = ColorStone600,
                modifier = Modifier
                    .fillMaxWidth()
                    .padding(bottom = 8.dp)
            )
        }

        Text(
            buildAnnotatedString {
                withStyle(style = SpanStyle(color = Color.Blue)) { append("Verifier") }
                append(" is requesting access to the following information")
            },
            fontFamily = Inter,
            fontWeight = FontWeight.Bold,
            fontSize = 20.sp,
            color = ColorStone950,
            modifier = Modifier
                .fillMaxWidth()
                .padding(bottom = 12.dp),
            textAlign = TextAlign.Center
        )

        Column(
            modifier = Modifier
                .fillMaxSize()
                .verticalScroll(rememberScrollState())
                .weight(weight = 1f, fill = false)
        ) {
            if (requestedFields.isEmpty() && !supportsSelectiveDisclosure) {
                allClaims.keys().asSequence().toList().sorted().forEach { claimName ->
                    Row(verticalAlignment = Alignment.CenterVertically) {
                        Checkbox(enabled = false, checked = true, onCheckedChange = { })
                        Text(
                            buildAnnotatedString {
                                withStyle(style = paragraphStyle) {
                                    append("\t\t")
                                    append(claimName.splitCamelCase().removeUnderscores())
                                }
                            },
                        )
                    }
                }
            } else {
                requestedFields.forEach {
                    Row(verticalAlignment = Alignment.CenterVertically) {
                        Checkbox(
                            enabled = supportsSelectiveDisclosure && !it.required(),
                            checked = selectedFields.contains(it.path()) || it.required() || !supportsSelectiveDisclosure,
                            onCheckedChange = { value ->
                                selectedFields = if (!value) {
                                    selectedFields.minus(it.path())
                                } else {
                                    selectedFields.plus(it.path())
                                }
                            }
                        )
                        Text(
                            buildAnnotatedString {
                                withStyle(style = paragraphStyle) {
                                    append("\t\t")
                                    append((it.name() ?: it.inputDescriptorId()).splitCamelCase().removeUnderscores())
                                }
                            },
                        )
                    }
                }
            }
        }

        Row(
            modifier = Modifier
                .fillMaxWidth()
                .padding(vertical = 12.dp)
                .navigationBarsPadding(),
            horizontalArrangement = Arrangement.spacedBy(8.dp),
        ) {
            Button(
                onClick = onCancel,
                shape = RoundedCornerShape(6.dp),
                colors = ButtonDefaults.buttonColors(
                    containerColor = Color.Transparent,
                    contentColor = ColorStone950,
                ),
                modifier = Modifier
                    .fillMaxWidth()
                    .border(1.dp, ColorStone300, RoundedCornerShape(6.dp))
                    .weight(1f)
            ) {
                Text("Cancel", fontFamily = Inter, fontWeight = FontWeight.SemiBold, color = ColorStone950)
            }

            Button(
                onClick = { onContinue(selectedFields) },
                shape = RoundedCornerShape(6.dp),
                colors = ButtonDefaults.buttonColors(containerColor = ColorEmerald900),
                modifier = Modifier
                    .fillMaxWidth()
                    .background(ColorEmerald900, RoundedCornerShape(6.dp))
                    .weight(1f)
            ) {
                Text(
                    text = if (hasMoreCredentials) "Next" else "Approve",
                    fontFamily = Inter,
                    fontWeight = FontWeight.SemiBold,
                    color = ColorBase50,
                )
            }
        }
    }
}

@Composable
fun Draft18CredentialSelector(
    requirements: List<Draft18CredentialRequirement>,
    credentialClaims: Map<String, JSONObject>,
    getRequestedFields: (Draft18PresentableCredential) -> List<Draft18RequestedField>,
    onContinue: (List<Draft18PresentableCredential>) -> Unit,
    onCancel: () -> Unit
) {
    var currentIndex by remember { mutableIntStateOf(0) }
    val selectedByRequirement =
        remember { mutableStateOf<Map<Int, Draft18PresentableCredential>>(emptyMap()) }

    val currentRequirement = requirements[currentIndex]
    val hasMoreRequirements = currentIndex + 1 < requirements.size

    fun selectCredential(credential: Draft18PresentableCredential) {
        val credentialId = credential.asParsedCredential().id()
        val current = selectedByRequirement.value[currentIndex]
        selectedByRequirement.value = if (current != null && current.asParsedCredential().id() == credentialId) {
            selectedByRequirement.value - currentIndex
        } else {
            selectedByRequirement.value + (currentIndex to credential)
        }
    }

    fun isSelected(credential: Draft18PresentableCredential): Boolean {
        return selectedByRequirement.value[currentIndex]?.asParsedCredential()?.id() ==
            credential.asParsedCredential().id()
    }

    fun getCredentialTitle(credential: Draft18PresentableCredential): String {
        val parsedCredential = credential.asParsedCredential()

        try {
            credentialClaims[parsedCredential.id()]?.getString("name")
                ?.takeIf { it.isNotBlank() }?.let { return it }
        } catch (_: Exception) {
        }

        try {
            credentialClaims[parsedCredential.id()]?.getJSONArray("type")?.let {
                for (i in 0 until it.length()) {
                    if (it.get(i).toString() != "VerifiableCredential") {
                        return it.get(i).toString().splitCamelCase()
                    }
                }
            }
        } catch (_: Exception) {
        }

        try {
            parsedCredential.asMsoMdoc()?.let { return credentialTypeDisplayName(it.doctype()) }
        } catch (_: Exception) {
        }

        try {
            parsedCredential.asDcSdJwt()?.let { return credentialTypeDisplayName(it.vct()) }
        } catch (_: Exception) {
        }

        return currentRequirement.displayName
    }

    fun getSelectedCredentials(): List<Draft18PresentableCredential> {
        return requirements.indices.mapNotNull { selectedByRequirement.value[it] }
    }

    Column(
        modifier = Modifier
            .fillMaxWidth()
            .padding(horizontal = 24.dp)
            .padding(top = 48.dp)
    ) {
        if (requirements.size > 1) {
            Text(
                text = "Requirement ${currentIndex + 1} of ${requirements.size}",
                fontFamily = Inter,
                fontWeight = FontWeight.Medium,
                fontSize = 14.sp,
                color = ColorStone600,
                modifier = Modifier
                    .fillMaxWidth()
                    .padding(bottom = 8.dp)
            )
        }

        Column(
            modifier = Modifier.fillMaxWidth(),
            horizontalAlignment = Alignment.CenterHorizontally
        ) {
            Text(
                text = "Select a credential for",
                fontFamily = Inter,
                fontWeight = FontWeight.Normal,
                fontSize = 16.sp,
                color = ColorStone600
            )
            Text(
                text = currentRequirement.displayName,
                fontFamily = Inter,
                fontWeight = FontWeight.Bold,
                fontSize = 20.sp,
                color = ColorBlue600
            )
        }

        Column(
            modifier = Modifier
                .fillMaxSize()
                .verticalScroll(rememberScrollState())
                .weight(weight = 1f, fill = false)
                .padding(top = 12.dp)
        ) {
            currentRequirement.credentials.forEach { credential ->
                Draft18CredentialSelectorItem(
                    credential = credential,
                    requestedFields = getRequestedFields(credential),
                    getCredentialTitle = { getCredentialTitle(it) },
                    isChecked = isSelected(credential),
                    onCheckedChange = { selectCredential(credential) }
                )
            }
        }

        Row(
            modifier = Modifier
                .fillMaxWidth()
                .padding(vertical = 12.dp)
                .navigationBarsPadding(),
            horizontalArrangement = Arrangement.spacedBy(8.dp),
        ) {
            Button(
                onClick = onCancel,
                shape = RoundedCornerShape(6.dp),
                colors = ButtonDefaults.buttonColors(
                    containerColor = Color.Transparent,
                    contentColor = ColorStone950,
                ),
                modifier = Modifier
                    .fillMaxWidth()
                    .border(1.dp, ColorStone300, RoundedCornerShape(6.dp))
                    .weight(1f)
            ) {
                Text("Cancel", fontFamily = Inter, fontWeight = FontWeight.SemiBold, color = ColorStone950)
            }

            Button(
                onClick = {
                    if (hasMoreRequirements) {
                        currentIndex += 1
                    } else {
                        onContinue(getSelectedCredentials())
                    }
                },
                shape = RoundedCornerShape(6.dp),
                colors = ButtonDefaults.buttonColors(containerColor = ColorStone600),
                modifier = Modifier
                    .fillMaxWidth()
                    .background(ColorStone600, RoundedCornerShape(6.dp))
                    .weight(1f)
            ) {
                Text(
                    text = if (hasMoreRequirements) "Next" else "Continue",
                    fontFamily = Inter,
                    fontWeight = FontWeight.SemiBold,
                    color = ColorBase50,
                )
            }
        }
    }
}

@Composable
fun Draft18CredentialSelectorItem(
    credential: Draft18PresentableCredential,
    requestedFields: List<Draft18RequestedField>,
    getCredentialTitle: (Draft18PresentableCredential) -> String,
    isChecked: Boolean,
    onCheckedChange: () -> Unit
) {
    var expanded by remember { mutableStateOf(false) }

    val bullet = "\u2022"
    val paragraphStyle = ParagraphStyle(textIndent = TextIndent(restLine = 12.sp))
    val displayFields = requestedFields.map { (it.name() ?: it.inputDescriptorId()).splitCamelCase().removeUnderscores() }

    Column(
        modifier = Modifier
            .fillMaxWidth()
            .padding(vertical = 8.dp)
            .border(1.dp, ColorBase300, RoundedCornerShape(8.dp))
    ) {
        Row(
            modifier = Modifier
                .fillMaxWidth()
                .padding(end = 8.dp)
                .padding(vertical = 8.dp),
            verticalAlignment = Alignment.CenterVertically
        ) {
            Checkbox(
                checked = isChecked,
                onCheckedChange = { onCheckedChange() },
                colors = CheckboxDefaults.colors(
                    checkedColor = ColorBlue600,
                    uncheckedColor = ColorStone300
                )
            )
            Text(
                text = getCredentialTitle(credential),
                fontFamily = Inter,
                fontWeight = FontWeight.SemiBold,
                fontSize = 18.sp,
                color = ColorStone950,
                modifier = Modifier.weight(1f)
            )
            if (expanded) {
                Image(
                    painter = painterResource(id = R.drawable.collapse),
                    contentDescription = stringResource(id = R.string.collapse),
                    modifier = Modifier.clickable { expanded = false }
                )
            } else {
                Image(
                    painter = painterResource(id = R.drawable.expand),
                    contentDescription = stringResource(id = R.string.expand),
                    modifier = Modifier.clickable { expanded = true }
                )
            }
        }

        if (expanded) {
            Text(
                buildAnnotatedString {
                    displayFields.forEach {
                        withStyle(style = paragraphStyle) {
                            append(bullet)
                            append("\t\t")
                            append(it)
                        }
                    }
                },
                modifier = Modifier.padding(16.dp)
            )
        }
    }
}

File: ios/Showcase/Targets/AppUIKit/Sources/wallet/Draft18OID4VPSupport.swift
import SpruceIDMobileSdk
import SpruceIDMobileSdkRs
import SwiftUI

class Draft18Signer: Draft18PresentationSigner {
    private let keyId: String
    private let jwkString: String
    private let didJwk = DidMethodUtils(method: DidMethod.jwk)

    init(keyId: String?) throws {
        self.keyId = keyId ?? DEFAULT_SIGNING_KEY_ID

        if !KeyManager.keyExists(id: self.keyId) {
            _ = KeyManager.generateSigningKey(id: self.keyId)
        }

        guard let jwk = KeyManager.getJwk(id: self.keyId) else {
            throw Oid4vpSignerError.illegalArgumentException(reason: "Invalid kid")
        }

        jwkString = jwk.description
    }

    func sign(payload: Data) async throws -> Data {
        guard let signature = KeyManager.signPayload(id: keyId, payload: [UInt8](payload)) else {
            throw Oid4vpSignerError.illegalArgumentException(reason: "Failed to sign payload")
        }

        return Data(signature)
    }

    func algorithm() -> String {
        let json = getGenericJSON(jsonString: jwkString)
        return json?.dictValue?["alg"]?.toString() ?? "ES256"
    }

    func verificationMethod() async -> String {
        try! await didJwk.vmFromJwk(jwk: jwkString)
    }

    func did() -> String {
        try! didJwk.didFromJwk(jwk: jwkString)
    }

    func cryptosuite() -> String {
        "ecdsa-rdfc-2019"
    }

    func jwk() -> String {
        jwkString
    }
}

struct Draft18CredentialRequirement {
    let descriptorId: String
    let displayName: String
    let credentials: [Draft18PresentableCredential]
}

struct Draft18DataFieldSelector: View {
    let requestedFields: [Draft18RequestedField]
    let selectedCredential: Draft18PresentableCredential
    let currentIndex: Int
    let totalCount: Int
    let onContinue: ([String]) -> Void
    let onCancel: () -> Void
    let allClaims: [String: GenericJSON]

    @State private var selectedFields: [String]
    let requiredFields: [String]
    let supportsSelectiveDisclosure: Bool

    init(
        requestedFields: [Draft18RequestedField],
        selectedCredential: Draft18PresentableCredential,
        currentIndex: Int = 0,
        totalCount: Int = 1,
        onContinue: @escaping ([String]) -> Void,
        onCancel: @escaping () -> Void,
        allClaims: [String: GenericJSON] = [:]
    ) {
        self.requestedFields = requestedFields
        self.selectedCredential = selectedCredential
        self.currentIndex = currentIndex
        self.totalCount = totalCount
        self.onContinue = onContinue
        self.onCancel = onCancel
        self.allClaims = allClaims
        supportsSelectiveDisclosure = selectedCredential.selectiveDisclosable()
        requiredFields = requestedFields.filter { $0.required() }.map { $0.path() }
        _selectedFields = State(initialValue: requiredFields)
    }

    func toggleBinding(for field: Draft18RequestedField) -> Binding<Bool> {
        Binding {
            selectedFields.contains(field.path())
        } set: { _ in
            if supportsSelectiveDisclosure && !field.required() {
                if selectedFields.contains(field.path()) {
                    selectedFields.removeAll(where: { $0 == field.path() })
                } else {
                    selectedFields.append(field.path())
                }
            }
        }
    }

    var hasMoreCredentials: Bool {
        currentIndex + 1 < totalCount
    }

    var body: some View {
        VStack {
            if totalCount > 1 {
                HStack {
                    Text("Credential \(currentIndex + 1) of \(totalCount)")
                        .font(.customFont(font: .inter, style: .medium, size: .p))
                        .foregroundStyle(Color("ColorStone500"))
                    Spacer()
                }
                .padding(.bottom, 8)
            }

            Group {
                Text("Verifier ")
                    .font(.customFont(font: .inter, style: .bold, size: .h2))
                    .foregroundColor(Color("ColorBlue600"))
                    + Text("is requesting access to the following information")
                    .font(.customFont(font: .inter, style: .bold, size: .h2))
                    .foregroundColor(Color("ColorStone950"))
            }
            .multilineTextAlignment(.center)

            ScrollView {
                if requestedFields.isEmpty && !supportsSelectiveDisclosure {
                    ForEach(Array(allClaims.keys.sorted()), id: \.self) { claimName in
                        SelectiveDisclosureItem(
                            fieldName: claimName,
                            required: true,
                            isChecked: .constant(true)
                        )
                    }
                } else {
                    ForEach(Array(requestedFields.enumerated()), id: \.offset) { _, field in
                        Draft18SelectiveDisclosureItem(
                            field: field,
                            required: field.required() || !supportsSelectiveDisclosure,
                            isChecked: toggleBinding(for: field)
                        )
                    }
                }
            }

            HStack {
                Button(action: onCancel) {
                    Text("Cancel")
                        .frame(maxWidth: .infinity)
                        .font(.customFont(font: .inter, style: .medium, size: .h4))
                }
                .foregroundColor(Color("ColorStone950"))
                .padding(.vertical, 13)
                .overlay(
                    RoundedRectangle(cornerRadius: 8)
                        .stroke(Color("ColorStone300"), lineWidth: 1)
                )

                Button {
                    onContinue(selectedFields)
                } label: {
                    Text(hasMoreCredentials ? "Next" : "Approve")
                        .frame(maxWidth: .infinity)
                        .font(.customFont(font: .inter, style: .medium, size: .h4))
                }
                .foregroundColor(.white)
                .padding(.vertical, 13)
                .background(Color("ColorEmerald900"))
                .clipShape(RoundedRectangle(cornerRadius: 8))
            }
            .fixedSize(horizontal: false, vertical: true)
        }
        .padding(.horizontal, 24)
        .navigationBarBackButtonHidden(true)
    }
}

struct Draft18SelectiveDisclosureItem: View {
    let field: Draft18RequestedField
    let required: Bool
    @Binding var isChecked: Bool

    private var displayName: String {
        field.name()?.camelCaseToWords().capitalized.replaceUnderscores()
            ?? field.inputDescriptorId().camelCaseToWords().capitalized.replaceUnderscores()
    }

    var body: some View {
        HStack {
            Toggle(isOn: $isChecked) {
                Text(displayName)
                    .font(.customFont(font: .inter, style: .regular, size: .h4))
                    .foregroundStyle(Color("ColorStone950"))
                    .frame(maxWidth: .infinity, alignment: .leading)
            }
            .toggleStyle(iOSCheckboxToggleStyle(enabled: !required))
        }
    }
}

struct Draft18CredentialSelector: View {
    let requirements: [Draft18CredentialRequirement]
    let credentialClaims: [String: [String: GenericJSON]]
    let getRequestedFields: (Draft18PresentableCredential) -> [Draft18RequestedField]
    let onContinue: ([Draft18PresentableCredential]) -> Void
    let onCancel: () -> Void

    @State private var currentIndex: Int = 0
    @State private var selectedByRequirement: [Int: Draft18PresentableCredential] = [:]

    var currentRequirement: Draft18CredentialRequirement {
        requirements[currentIndex]
    }

    var hasMoreRequirements: Bool {
        currentIndex + 1 < requirements.count
    }

    func selectCredential(_ credential: Draft18PresentableCredential) {
        let credentialId = credential.asParsedCredential().id()
        if let current = selectedByRequirement[currentIndex],
           current.asParsedCredential().id() == credentialId
        {
            selectedByRequirement.removeValue(forKey: currentIndex)
        } else {
            selectedByRequirement[currentIndex] = credential
        }
    }

    func isSelected(_ credential: Draft18PresentableCredential) -> Bool {
        guard let selected = selectedByRequirement[currentIndex] else {
            return false
        }

        return selected.asParsedCredential().id() == credential.asParsedCredential().id()
    }

    func toggleBinding(for credential: Draft18PresentableCredential) -> Binding<Bool> {
        Binding {
            isSelected(credential)
        } set: { _ in
            selectCredential(credential)
        }
    }

    func getSelectedCredentials() -> [Draft18PresentableCredential] {
        requirements.indices.compactMap { selectedByRequirement[$0] }
    }

    func getCredentialTitle(_ credential: Draft18PresentableCredential) -> String {
        let parsedCredential = credential.asParsedCredential()
        if let name = credentialClaims[parsedCredential.id()]?["name"]?.toString() {
            return name
        }

        if let types = credentialClaims[parsedCredential.id()]?["type"]?.arrayValue {
            for type in types where type.toString() != "VerifiableCredential" {
                return type.toString().camelCaseToWords()
            }
        }

        if let mdoc = parsedCredential.asMsoMdoc() {
            return credentialTypeDisplayName(for: mdoc.doctype())
        }

        if let dcSdJwt = parsedCredential.asDcSdJwt() {
            return credentialTypeDisplayName(for: dcSdJwt.vct())
        }

        return currentRequirement.displayName
    }

    func goToNextOrFinish() {
        if hasMoreRequirements {
            currentIndex += 1
        } else {
            onContinue(getSelectedCredentials())
        }
    }

    var body: some View {
        VStack {
            if requirements.count > 1 {
                HStack {
                    Text("Requirement \(currentIndex + 1) of \(requirements.count)")
                        .font(.customFont(font: .inter, style: .medium, size: .p))
                        .foregroundStyle(Color("ColorStone500"))
                    Spacer()
                }
                .padding(.bottom, 8)
            }

            VStack(spacing: 4) {
                Text("Select a credential for")
                    .font(.customFont(font: .inter, style: .regular, size: .h3))
                    .foregroundStyle(Color("ColorStone700"))

                Text(currentRequirement.displayName)
                    .font(.customFont(font: .inter, style: .bold, size: .h2))
                    .foregroundStyle(Color("ColorBlue600"))
            }
            .multilineTextAlignment(.center)
            .padding(.bottom, 8)

            ScrollView {
                ForEach(Array(currentRequirement.credentials.enumerated()), id: \.offset) { _, credential in
                    Draft18CredentialSelectorItem(
                        credential: credential,
                        requestedFields: getRequestedFields(credential),
                        getCredentialTitle: getCredentialTitle,
                        isChecked: toggleBinding(for: credential)
                    )
                }
            }

            HStack {
                Button(action: onCancel) {
                    Text("Cancel")
                        .frame(maxWidth: .infinity)
                        .font(.customFont(font: .inter, style: .medium, size: .h4))
                }
                .foregroundColor(Color("ColorStone950"))
                .padding(.vertical, 13)
                .overlay(
                    RoundedRectangle(cornerRadius: 8)
                        .stroke(Color("ColorStone300"), lineWidth: 1)
                )

                Button {
                    goToNextOrFinish()
                } label: {
                    Text(hasMoreRequirements ? "Next" : "Continue")
                        .frame(maxWidth: .infinity)
                        .font(.customFont(font: .inter, style: .medium, size: .h4))
                }
                .foregroundColor(.white)
                .padding(.vertical, 13)
                .background(Color("ColorStone600"))
                .clipShape(RoundedRectangle(cornerRadius: 8))
            }
            .fixedSize(horizontal: false, vertical: true)
        }
        .padding(.horizontal, 24)
        .navigationBarBackButtonHidden(true)
    }
}

struct Draft18CredentialSelectorItem: View {
    let credential: Draft18PresentableCredential
    let requestedFields: [String]
    let getCredentialTitle: (Draft18PresentableCredential) -> String
    @Binding var isChecked: Bool

    @State private var expanded = false

    init(
        credential: Draft18PresentableCredential,
        requestedFields: [Draft18RequestedField],
        getCredentialTitle: @escaping (Draft18PresentableCredential) -> String,
        isChecked: Binding<Bool>
    ) {
        self.credential = credential
        self.requestedFields = requestedFields.map {
            ($0.name() ?? $0.inputDescriptorId()).capitalized
        }
        self.getCredentialTitle = getCredentialTitle
        _isChecked = isChecked
    }

    var body: some View {
        VStack {
            HStack {
                Toggle(isOn: $isChecked) {
                    Text(getCredentialTitle(credential))
                        .font(.customFont(font: .inter, style: .semiBold, size: .h3))
                        .foregroundStyle(Color("ColorStone950"))
                }
                .toggleStyle(iOSCheckboxToggleStyle())
                Spacer()
                if expanded {
                    Image("Collapse")
                        .onTapGesture { expanded = false }
                } else {
                    Image("Expand")
                        .onTapGesture { expanded = true }
                }
            }

            VStack(alignment: .leading) {
                ForEach(requestedFields, id: \.self) { field in
                    Text("• \(field)")
                        .font(.customFont(font: .inter, style: .regular, size: .h4))
                        .foregroundStyle(Color("ColorStone950"))
                        .frame(maxWidth: .infinity, alignment: .leading)
                }
            }
            .hide(if: !expanded)
        }
        .padding(16)
        .overlay(
            RoundedRectangle(cornerRadius: 8)
                .stroke(Color("ColorBase300"), lineWidth: 1)
        )
        .padding(.vertical, 6)
    }
}
